src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MaxLengthAttribute.cs

```diff
@@ -76,11 +78,16 @@ public override bool IsValid(object value)
             {
                 length = str.Length;
             }
+            else if (value is ICollection collection)
+            {
+                length = collection.Count;
+            }
             else
             {
-                if (value is ICollection collection)
+                Type genericCol = value.GetType().GetInterfaces().FirstOrDefault(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(ICollection<>));
```

Comment:
Stepping back a bit, it seems that it won't be needed in the majority of scenarios we are trying to enable. It turns out that most concrete types that implement `ICollection<T>` also implement the non-generic `ICollection`, so they will match the previous condition.

To be honest, there is the fact that code smells (it seems correct but it looks more complex and expensive than I was hoping it would be) and if we are going to include it we should probably move it into an internal helper so that both attributes can share it. 

cc @ajcvickers @lajones

Reply:
@divega I'm ok with just targeting `ICollection<T>`, anyway please also take a look at #23578.

Reply:
@weitzhandler I assume you mean targeting `ICollection` (non-generic). Then fine, let's remove the whole check of `ICollection<T>`.  

Re `ICountable`, I don't hate the idea :smile: I am not sure what the best decision is, but if you want to make the case for it more complete, I suggest you show right there in the top comment the in the issue the code you have to write in order to obtain the `Count` of and `ICollection<T>` without knowing what the `T` is.

Reply:
OK. Will mess with it later when I find some time.
Anyway, I think we should leave the `ICollection` check here, so if the value does implement `ICollection` (most generic collections do), we will spare the reflection overhead.
BTW, `HashSet<T>` is an example of a collection not inheriting from `ICollection`, and is pretty common use, also in validation. I wouldn't have bothered implementing this just for the heck of it...

Reply:
@divega 
I think we should leave the `ICollection` check here, so if the value does implement `ICollection` (most collections do), we will spare the reflection overhead.
BTW, `HashSet<T>` is an example of a collection not inheriting from `ICollection`, and is pretty common use, also in validation. I wouldn't have bothered implementing this just for the heck of it...

Anyway, I updated the post on the `ICountable`, please check it out #23578.

Reply:
@weitzhandler 

> I think we should leave the ICollection check here, so if the value does implement ICollection (most collections do), we will spare the reflection overhead.

:+1: that is what I meant as well.

> HashSet<T> is an example of a collection not inheriting from ICollection

You are correct. I missed `HashSet<T>` when I looked around, and that one is quite important.

I talked about this briefly with @ajcvickers and realized that when we wrote the comment at https://github.com/dotnet/corefx/issues/23461#issuecomment-324767348 we both were thinking that the way to go was to just use reflection to look for a `Count` property that has a getter and returns int, i.e. without all the goop to try to find the "right" interface. 

Incidentally, I believe this simple pattern matching aligns better with the spirit of your `ICountable` proposal.